common/scala/src/main/scala/whisk/core/entity/Exec.scala

```diff
@@ -174,60 +198,66 @@ protected[core] object Exec
+                        } else None
                     }
-                    SequenceExec(comp)
-
-                case Exec.BLACKBOX =>
-                    val image: String = obj.getFields("image") match {
-                        case Seq(JsString(i)) => i
-                        case _                => throw new DeserializationException(s"'image' must be a string defined in 'exec' for '${Exec.BLACKBOX}' actions")
-                    }
-                    val code: Option[String] = obj.getFields("code") match {
-                        case Seq(JsString(i)) => if (i.trim.nonEmpty) Some(i) else None
-                        case Seq(_)           => throw new DeserializationException(s"if defined, 'code' must a string defined in 'exec' for '${Exec.BLACKBOX}' actions")
-                        case _                => None
+                    CodeExecAsAttachment(manifest, jar, main)
+                }.getOrElse {
+                    val code: String = obj.fields.get("code") match {
+                        case Some(JsString(c)) => c
+                        case _                 => throw new DeserializationException(s"'code' must be a string defined in 'exec' for '$kind' actions")
                     }
-                    BlackBoxExec(image, code, optMainField)
-
-                case _ => throw new DeserializationException(s"kind '$kind' not in $runtimes")
+                    CodeExecAsString(manifest, code, optMainField)
+                }
+            } else if (kind == Exec.SEQUENCE) {
+                val comp: Vector[FullyQualifiedEntityName] = obj.getFields("components") match {
+                    case Seq(JsArray(components)) => components map { FullyQualifiedEntityName.serdes.read(_) }
+                    case Seq(_)                   => throw new DeserializationException(s"'components' must be an array")
+                    case _                        => throw new DeserializationException(s"'components' must be defined for sequence kind")
+                }
+                SequenceExec(comp)
+            } else if (kind == Exec.BLACKBOX) {
+                val image: String = obj.getFields("image") match {
+                    case Seq(JsString(i)) => i
+                    case _                => throw new DeserializationException(s"'image' must be a string defined in 'exec' for '${Exec.BLACKBOX}' actions")
+                }
+                val code: Option[String] = obj.getFields("code") match {
+                    case Seq(JsString(i)) => if (i.trim.nonEmpty) Some(i) else None
+                    case Seq(_)           => throw new DeserializationException(s"if defined, 'code' must a string defined in 'exec' for '${Exec.BLACKBOX}' actions")
+                    case _                => None
+                }
+                BlackBoxExec(image, code, optMainField)
+            } else {
+                throw new DeserializationException(s"kind '$kind' not in $runtimes")
```

Comment:
I believe this is dead code. Rewrite the above to a `match` on kind would make this obvious I think.